### ISSUE TITLE
Possible issues

### ISSUE THREAD

**Issue by zulihit** [backprop07/Self-Certainty#7]
Hi, thanks for releasing the Self-Certainty code and the paper "Scalable Best-of-N Selection for Large Language Models via Self-Certainty". I have a few questions:

1）After adding a pad token, the code does:

tokenizer.add_special_tokens({"pad_token": "<pad>"})
llm.config.pad_token_id = tokenizer.pad_token_id
llm.resize_token_embeddings(len(tokenizer))
llm.embed_tokens = torch.nn.Embedding(
    llm.config.vocab_size, llm.config.hidden_size, padding_idx=llm.config.pad_token_id
).to(device)

My understanding is that resize_token_embeddings already preserves pretrained embeddings and only initializes new rows, while reassigning llm.embed_tokens here will reinitialize all embeddings and discard the original weights. Is this intended, or should the reassignment of llm.embed_tokens be removed?

2）Inside @torch.no_grad() we also call torch.set_grad_enabled(False). Since the decorator already disables grad, is this just leftover / redundant code that can be removed?

3）In the length grouping:
for idx, text in enumerate(outputs):
    if len(text) > 3 * 1024:
        groups["large"]["outputs"].append(text)
        groups["large"]["indices"].append(idx)
    elif len(text) > 6 * 1024:
        groups["medium"]["outputs"].append(text)
        groups["medium"]["indices"].append(idx)
    else:
        groups["small"]["outputs"].append(text)
        groups["small"]["indices"].append(idx)

Because len(text) > 6*1024 also satisfies len(text) > 3*1024, the elif branch seems unreachable and medium is always empty. Should the thresholds be reversed (first check > 6*1024, then > 3*1024)?

4）For logits alignment, the code uses:
logits = llm(batch_ids, attention_mask=batch_attention_mask).logits
logits = logits[:, input_length:, :]
logp = F.log_softmax(logits, dim=-1)
logprob_sum = logp.sum(dim=-1)
confidence_list = confidence_logprob_sum(logprob_sum, output_attention_mask, V)

For a causal LM where logits[:, t, :] predicts token at position t+1, if input_length is the prompt length, it seems more standard to use logits[:, input_length - 1:-1, :] so that we use the positions that predict the output tokens and drop the final extra step. Is the current logits[:, input_length:, :] slicing intentional or an off-by-one issue?
